src/MongoDB.Analyzer/Core/Builders/BuildersResolveVariables.cs

```diff
@@ -0,0 +1,393 @@
+            var binaryExpression = SyntaxFactory.BinaryExpression(syntaxKind, LHS as ExpressionSyntax, rightSide);
+            rewrittenBuildersExpression = SyntaxFactory.AssignmentExpression(SyntaxKind.SimpleAssignmentExpression, variableIdentifier, binaryExpression);
+        }
+
+        return (rewrittenBuildersExpression, constantMappers, argumentTypeName);
+    }
+
+    private static ExpressionAnalysisContext GenerateExpressionAnalysisContext(
+        SyntaxNode syntaxNode,
+        List<SyntaxNode> childBuilderNodes,
+        ContextInformation contextInformation)
+    {
+        var (rewrittenBuildersExpression, constantMappers, argumentTypeName) = RewriteBuildersExpression(syntaxNode, childBuilderNodes, contextInformation);
+
+        if (rewrittenBuildersExpression == null)
+        {
+            return null;
+        }
+
+        var (finalBuildersExpression, finalConstantMapper) = ConstantsMapperComposer.Compose(rewrittenBuildersExpression, constantMappers);
+
+        if (finalBuildersExpression is VariableDeclaratorSyntax variableDeclaratorSyntax)
+        {
+            finalBuildersExpression = variableDeclaratorSyntax.Initializer.Value;
+        }
+        else if (finalBuildersExpression is AssignmentExpressionSyntax)
+        {
+            while (finalBuildersExpression is AssignmentExpressionSyntax assignmentExpression)
+            {
+                finalBuildersExpression = assignmentExpression.Right;
+            }
+        }
+
+        var expressionAnalysisContext = new ExpressionAnalysisContext(new ExpressionAnalysisNode(syntaxNode, argumentTypeName, finalBuildersExpression, finalConstantMapper, new Location[] { }));
+        return expressionAnalysisContext;
+    }
+
+    private static void ProcessDeclaration(VariableDeclaratorSyntax variableDeclarator, ContextInformation contextInformation, HashSet<string> dirtyVariables)
+    {
+        var childBuilderNodes = new List<SyntaxNode>();
+        var declarationExpression = variableDeclarator.Initializer.Value;
+        var canEvaluate = CanEvaluate(declarationExpression, childBuilderNodes, contextInformation);
+
+        var variableNames = new List<string>();
+        variableNames.Add(variableDeclarator.Identifier.ValueText);
+
+        var variableValues = contextInformation.VariableValues;
+        var expressionAnalysisContexts = contextInformation.ExpressionAnalysisContexts;
+
+        if (canEvaluate)
+        {
+            var expressionAnalysisContext = GenerateExpressionAnalysisContext(variableDeclarator, childBuilderNodes, contextInformation);
+
+            foreach (var variableName in variableNames)
+            {
+                var variableValue = new VariableValue(expressionAnalysisContext, new List<Location>());
+                dirtyVariables.Add(variableName);
+
+                if (!variableValues.ContainsKey(variableName))
```

Comment:
variableValues.TryGet
That way we are doing a single lookup, instead of two with ContainsKey and variableValues[variableName].